kafka/server|hashing: bazelize benchmarks

IoannisRP opened the pull request:
implements:
- [CORE-7484](https://redpandadata.atlassian.net/browse/CORE-7484): `hashing/tests/hash_bench.cc`
- [CORE-7502](https://redpandadata.atlassian.net/browse/CORE-7502): `kafka/server/tests/fetch_bench.cc`
- [CORE-7503](https://redpandadata.atlassian.net/browse/CORE-7503): `kafka/server/tests/fetch_plan_bench.cc`
- [CORE-7521](https://redpandadata.atlassian.net/browse/CORE-7521): `kafka/server/tests/produce_partition_bench.cc`
- [CORE-7522](https://redpandadata.atlassian.net/browse/CORE-7522): `kafka/server/tests/quota_manager_bench.cc`

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none


[CORE-7484]: https://redpandadata.atlassian.net/browse/CORE-7484?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CORE-7502]: https://redpandadata.atlassian.net/browse/CORE-7502?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CORE-7503]: https://redpandadata.atlassian.net/browse/CORE-7503?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CORE-7521]: https://redpandadata.atlassian.net/browse/CORE-7521?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CORE-7522]: https://redpandadata.atlassian.net/browse/CORE-7522?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

IoannisRP committed:
hashing: bazelize hashing_bench

IoannisRP committed:
kafka/server: bazelize fetch_bench

IoannisRP committed:
kafka/server: bazelize fetch_plan_bench

IoannisRP committed:
kafka/server: bazelize produce_partition_bench

IoannisRP committed:
kafka/server: bazelize quota_manager_bench

rockwotj commented:
FYI I'm also working on some of the benchmarks too

IoannisRP commented:
@rockwotj  these are all the benchmarks i had... it is a draft just because it depends on #24792.
Do we have any overlapping?

rockwotj commented:
Just hashing, it was an accident, I dropped my version and added you as a reviewer.

rockwotj commented on src/v/hashing/tests/BUILD:
```diff
@@ -24,3 +24,21 @@ redpanda_cc_btest_no_seastar(
 )
+
+redpanda_cc_bench(
+    name = "hashing_bench",
```
we need the binary names to match the CMake binaries for some reporting to work. This needs to be `hashing_bench_rpbench`

rockwotj commented on src/v/kafka/server/tests/BUILD:
```diff
@@ -749,3 +749,101 @@ redpanda_cc_gtest(
 )
+
+redpanda_cc_bench(
+    name = "fetch_bench",
```
`kafka_fetch_rpbench`

rockwotj commented on src/v/kafka/server/tests/BUILD:
```diff
@@ -749,3 +749,101 @@ redpanda_cc_gtest(
+    srcs = [
+        "fetch_bench.cc",
+    ],
+    args = [
+        "--duration=1",
+        "--runs=1",
+    ],
```
Please remove this. I'll be adding a test target for benchmarks that does this automatically soon, but we don't want normal benchmarks to run like this :)

rockwotj commented on src/v/kafka/server/tests/BUILD:
```diff
@@ -749,3 +749,101 @@ redpanda_cc_gtest(
+    srcs = [
+        "fetch_plan_bench.cc",
+    ],
+    args = [
+        "--duration=1",
+        "--runs=1",
+    ],
```
ditto

rockwotj commented on src/v/kafka/server/tests/BUILD:
```diff
@@ -749,3 +749,101 @@ redpanda_cc_gtest(
+)
+
+redpanda_cc_bench(
+    name = "fetch_plan_bench",
```
`kafka_fetch_plan_rpbench`

rockwotj commented on src/v/kafka/server/tests/BUILD:
```diff
@@ -749,3 +749,101 @@ redpanda_cc_gtest(
+)
+
+redpanda_cc_bench(
+    name = "produce_partition_bench",
```
`kafka_produce_partition_rpbench`

rockwotj commented on src/v/kafka/server/tests/BUILD:
```diff
@@ -749,3 +749,101 @@ redpanda_cc_gtest(
+    srcs = [
+        "produce_partition_bench.cc",
+    ],
+    args = [
+        "--duration=1",
+        "--runs=1",
+    ],
```
please remove

rockwotj commented on src/v/kafka/server/tests/BUILD:
```diff
@@ -749,3 +749,101 @@ redpanda_cc_gtest(
+)
+
+redpanda_cc_bench(
+    name = "quota_manager_bench",
```
`quota_manager_rpbench`

rockwotj commented on src/v/kafka/server/tests/BUILD:
```diff
@@ -749,3 +749,101 @@ redpanda_cc_gtest(
+    srcs = [
+        "quota_manager_bench.cc",
+    ],
+    args = [
+        "--duration=1",
+        "--runs=1",
+    ],
```
please remove

rockwotj reviewed (commented):
LGTM thanks! Just need to rename some targets and we'll be good to go

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/60719-->
<!--38f74414-8951-41d7-a973-88db58bb2e63-->
<!--title-->
## Retry command for [Build#60719](https://buildkite.com/redpanda/redpanda/builds/60719)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/timequery_test.py::TimeQueryKafkaTest.test_timequery_below_start_offset
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--81996e0b-a8d3-4adf-b152-8c67bab93466-->
<!--TEST RESULTS-->
<!--c561d2f9-1df3-4ebc-a6de-3dd870965c03-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60719">build#60719</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60719#019465d1-4c76-42f1-9594-19772fdf84e6 | FLAKY | 1/2 |
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_reassignments_kafka_cli | ducktape | https://buildkite.com/redpanda/redpanda/builds/60719#01946617-af00-4c63-abd4-114a443a7f76 | FLAKY | 2/6 |
| rptest.tests.timequery_test.TimeQueryKafkaTest.test_timequery_below_start_offset | ducktape | https://buildkite.com/redpanda/redpanda/builds/60719#0194662a-3525-4585-8134-49dfc7ec2dbd | FAIL | 0/1 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60844">build#60844</a></summary>


| test_id                                                                                                                                              | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.audit_log_test.AuditLogTestsAppLifecycle.test_app_lifecycle                                                                             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/60844#01946fdc-dd78-42e7-842e-a10e7fe1b58f | FLAKY         | 5/6      |
| rptest.tests.delete_records_test.DeleteRecordsTest.test_delete_records_concurrent_truncations.cloud_storage_enabled=True.truncate_point=start_offset | ducktape    | https://buildkite.com/redpanda/redpanda/builds/60844#01947003-cf91-407e-8f50-73c249086e05 | FLAKY         | 5/6      |
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_reassignments_kafka_cli                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/60844#01947003-cf8f-4a18-8c7d-b30bfc3dc6cb | FLAKY         | 4/6      |
</details>
<!--/body-->

IoannisRP commented:
force-push changes:
- renamed benchmark targets to match cmake names

BenPope commented on src/v/kafka/server/tests/BUILD:
```diff
@@ -749,3 +749,28 @@ redpanda_cc_gtest(
+    ],
+    cpu = 2,
+    memory = "4GiB",
+    tags = ["exclusive"],
```
@rockwotj would this be covered by the test target for benchmarks?

rockwotj replied:
Yes

BenPope commented on src/v/kafka/server/tests/fetch_bench.cc:
```diff
@@ -173,12 +162,14 @@ struct fetch_bench_fixture : redpanda_thread_fixture {
                 co_await producer.produce_to_partition(
-                  t, model::partition_id(pid), {std::move(msg)});
+                  t,
+                  model::partition_id(static_cast<int>(pid)),
```
nitpick: An alternative is to accept `int32_t total_partition_count` and make the loop variable a `model::partition_id`

BenPope reviewed (dismissed):
In general, we prefer sized types. I.e., instead of `int`, use `int32_t`, the type of `model::partition_id::type`.

IoannisRP commented:
force push:
- rebased to dev

IoannisRP committed:
bazel: add duration and runs arguments to benchmark target

IoannisRP commented:
change in [force-push](https://github.com/redpanda-data/redpanda/compare/b7411f931ae087b2a0af0bf6589aea4e35b457b8..cd14752f76268339f1018883a52dd64d8d7268e1):
- added `runs` & `duration` arguments to `redpanda_cc_bench`
- changed static casts from `int`->`int32_t`

IoannisRP commented:
TODO: fix `kafka_rpbench_test`

It currently fails in debug mode. I need to disable it in that mode.

IoannisRP committed:
bazel: enable constraints for benchmark targets

- Enable benchmark constraints for smoke test
- Add config setting for optimised builds

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/cd14752f76268339f1018883a52dd64d8d7268e1..5d9f408df9509d2eebe6317d3e33e4c8cba48a0d):
- add constrains to benchmark rule
- add config setting to detect optimised build
- disable kafka_fetch_rpbench_test for non-optimised builds
